ruby/ext/google/protobuf_c/encode_decode.c

```diff
@@ -613,6 +688,23 @@ static void add_handlers_for_oneof_field(upb_handlers *h,
   upb_handlerattr_uninit(&attr);
 }
 
+static bool add_unknown_handler(void* closure, const void* hd, const char* buf,
+                         size_t size) {
+  encodeunknown_handlerfunc handler =
+      ((unknownfields_handlerdata_t*)hd)->handler;
+
+  MessageHeader* msg = (MessageHeader*)closure;
+  stringsink* unknown = DEREF(Message_data(msg), 0, stringsink*);
```

Comment:
Seems like you should add a stringsink* to `MessageHeader` instead of doing all this casting/DEREF?

If that doesn't work, at least add get/set functions for the stringsink.